docs/advanced_queries.rst

```diff
@@ -0,0 +1,206 @@
+and matching the main entity name in the query with the method name.
+
+For example here we query using a subquery, but the main entity name is `nodes` and the method name is the same.
+Note that the parameter to pass PQL query name is ``pql``.
+
+.. code-block:: python
+
+   >>> nodes = db.nodes(pql="""
+   >>>     nodes {
+   >>>       facts {
+   >>>         name = "operatingsystem" and
+   >>>         value = "Debian"
+   >>>       }
+   >>>     }
+   >>> """
+   >>> for node in nodes:
+   >>>     print(node.name)
+   foo.example.com
+   bar.example.com
+
+Note that you CANNOT use projections (fields/functions) in queries using these methods. If you want to use them,
+use the raw PQL query instead (below).
+
+
+Raw dicts with PQL query
+------------------------
+
+If you want to use projections (fields/functions) in your PQL query or you just prefer to get plain dicts instead
+of rich types, use ``pql()`` method like below.
+
+For example to get just a list of certnames of nodes that have the fact "operatingsystem"
+set to "Debian":
+
+.. code-block:: python
+
+   >>> nodes = db.pql("""
+   >>>     nodes[certname] {
+   >>>       facts {
+   >>>         name = "operatingsystem" and
+   >>>         value = "Debian"
+   >>>       }
+   >>>     }
+   >>> """
+   >>> for node in nodes:
+   >>>     print(node)
+   {'certname': 'foo.example.com'}
+   {'certname': 'bar.example.com'}
+
+Please see `the official PQL documentation <https://puppet.com/docs/puppetdb/latest/api/query/v4/pql.html>`_
+for more examples .
+
+
+AST query
+---------
+
+You can pass the AST query in the ``query`` parameter of the methods supporting this type of queries.
+
+An example of the same query as above but rewritten in AST:
+
+.. code-block:: python
+
+   >>> nodes = db.nodes(query="""
+   >>>     ["subquery", "facts",
+   >>>       ["and",
+   >>>         ["=", "name", "operatingsystem"],
+   >>>         ["=", "value", "Debian"]
+   >>>       ]
+   >>>     ]
+   >>> """
+   >>> for node in nodes:
+   >>>     print(node.name)
+   foo.example.com
+   bar.example.com
+
+Please see `the official AST documentation <https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html>`_
+for more examples .
+
+AST QueryBuilder
+^^^^^^^^^^^^^^^^
+
+pypuppetdb comes shipped also with a ``QueryBuilder`` component that, as the name suggests,
+allows users to build PuppetDB AST queries in an Object-Oriented fashion.
+Vastly superior to constructing long strings than adding additional clauses to fulfill
+new requirements.
+
+Above example rewritten using it:
+
+.. code-block:: python
+
+   >>> from pypuppetdb.QueryBuilder import *
+   >>> op = AndOperator()
+   >>> op.add(EqualsOperator('name', 'operatingsystem'))
+   >>> op.add(EqualsOperator('value', 'Debian'))
+   >>> query = SubqueryOperator('facts')
+   >>> query.add_query(op)
+   >>> print(query)
+   >>> # TODO: fix above code as this is not the query we wanted :/
```

Comment:
I need some help here as I can't recreate the AST query from the section above using the QueryBuilder syntax... (Maybe that's the reason I am working on this PQL PR..? ;P )

Reply:
I searched the code and as far as i can tell the query builder never uses subquery directly and uses explicit subqueries using select_*.  I think the folllowing may represent the same query

```python
    >>> op = InOperator('certname')
    >>> query = ExtractOperator()
    >>> query.add_field(str('certname'))
    >>> subquery = SubqueryOperator('facts')
    >>> _add = AndOperator()
    >>> _add.add(EqualsOperator('name', 'operatingsystem'))
    >>> _add.add(EqualsOperator('value', 'Debian'))
    >>> subquery.add_query(_add)
    >>> query.add_query(subquery)
    >>> op.add_query(query)
    >>> print(op)
   ["in", "certname", ["extract", ["certname"], ["select_facts", ["and", ["=", "name", "operatingsystem"], ["=", "value", "Debian"]]]]]
```

Reply:
Great, thanks!